packages/cube/pipeline/cube-validation.ttl

```diff
@@ -0,0 +1,229 @@
+        [ shacl:report (_:getProfile) ]
+        [ base:flatten () ]
+        [ ntriples:serialize () ]
+        _:stdout
+      )
+    ]
+.
+
+<check-cube-observations> a p:Pipeline ;
+  p:variables [ p:variable _:constraint , _:batchSize , _:maxViolations ] ;
+  p:steps
+    [
+      p:stepList ( 
+        _:stdin
+        [ n3:parse () ]
+        _:sortBySubject
+        _:toDatasetBySubject
+        _:batch
+        [ shacl:report 
+          [code:name "shape" ; code:value _:getConstraint ] ,
+          [code:name "maxErrors" ; code:value "maxViolations"^^p:VariableName ] 
+        ]
+        [ base:flatten () ]
+        [ ntriples:serialize () ]
+        _:stdout
+      )
+    ]
+.
+
+# relying on cbd is not ideal (vendor specific)
+_:queryConstraint sparql:construct
+    [ code:name "endpoint"; code:value "endpoint"^^p:VariableName ] , 
+    [
+      code:name "query";
+      code:value """
+      #pragma describe.strategy cbd
+
+      PREFIX cube: <https://cube.link/>
+
+      DESCRIBE ?s
+      WHERE {
+        <${cube}> cube:observationConstraint ?s .
+      }
+      """^^code:EcmaScriptTemplateLiteral 
+    ]
+.
+
+_:queryObservations sparql:construct
+    [
+      code:name "endpoint";
+      code:value "endpoint"^^p:VariableName
+    ] , 
+    [
+      code:name "query";
+      code:value """
+      PREFIX cube: <https://cube.link/>
+
+      CONSTRUCT { ?s ?p ?o }
+      WHERE {
+        <${cube}> cube:observationSet/cube:observation ?s .
+        ?s ?p ?o 
+      }
+      """^^code:EcmaScriptTemplateLiteral 
+    ]
+.
+
+_:stdin a p:Step ;
+  code:implementedBy "() => process.stdin"^^code:EcmaScript  
+.
+
+_:stdout a p:Step ;
+  code:implementedBy "() => process.stdout"^^code:EcmaScript  
+.
+
+_:toDataset a p:Step ;
+	code:implementedBy 
+    [
+      rdf:type code:EcmaScript ;
+      code:link <node:rdf-stream-to-dataset-stream/toDatasetStream.js> ;
+    ] 
+.
+
+_:toDatasetBySubject a p:Step ;
+	code:implementedBy 
+    [
+      rdf:type code:EcmaScript ;
+      code:link <node:rdf-stream-to-dataset-stream/bySubject.js> ;
+    ] 
+. 
+
+_:batch a p:Step ;
+  code:implementedBy 
+    [ 
+      a code:EcmaScriptModule ;
+      code:link <file:../lib/batch.js#batch> ;
+    ] ;
+  code:arguments ("batchSize"^^p:VariableName)
+.
+
+_:getProfile a p:Pipeline , p:ReadableObjectMode;
+  p:steps
+    [
+      p:stepList
+        (
+          [ http:get [ code:name "url" ; code:value "profile"^^p:VariableName ] ] 
+          [ n3:parse () ]
+        ) 
+    ] 
+  .
+
+_:getConstraint a p:Pipeline , p:ReadableObjectMode ;
+  p:variables [ p:variable _:constraint ] ;
+  p:steps
+    [
+      p:stepList
+        (
+          _:readConstraint
+          _:toDataset
+          _:addTarget
+          [ base:flatten () ]
+        ) 
+    ] 
+.
+
+_:readConstraint a p:Step ;
+  code:implementedBy 
+    [ 
+      a code:EcmaScriptModule ;
+      code:link <node:rdf-utils-fs#fromFile> ;
+    ] ;
+  code:arguments ("constraint"^^p:VariableName)
+.
```

Comment:
I added a suggestion to remove `rdf-utils-fs` because `@zazuko/env-node` provides that functionality. 

This should work

```suggestion
_:readConstraint a p:Step ;
  code:implementedBy """
  function (path) {
    return this.env.fromFile(path)
  }
  """^^code:EcmaScript ;
  code:arguments ("constraint"^^p:VariableName)
.
```

Reply:
I will try